src/penn_chime/validators/validators.py

```diff
@@ -0,0 +1,69 @@
+"""design pattern via https://youtu.be/S_ipdVNSFlo?t=2153"""
+
+from typing import Optional
+from datetime import date, datetime
+
+from .base import Validator
+
+
+class Bounded(Validator):
+    """A bounded number."""
+    def __init__(
+            self,
+            lower_bound: Optional[float] = None,
+            upper_bound: Optional[float] = None) -> None:
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.message = {
+            (lower_bound, upper_bound): f"in ({self.lower_bound}, {self.upper_bound})",
+            (None, upper_bound): f"less than {self.upper_bound}",
+            (lower_bound, None): f"greater than {self.lower_bound}",
+            (None, None): "ACTUALLY the value is unbounded"
```

Comment:
I don't think this initialization is doing what it intends, and Python is being too lenient to tell you. In the case where one or both of those bounds is `None`, have a look at how many entries actually appear in the dictionary.

Reply:
This is written as if it were pattern matching, but it's not - those keys have particular values, some of which overlap.

Reply:
hm. i think to be sure we could put the first three cases in `{}` dict initialization and then follow it with ```self.message[(None, None)] = "the vacuous case message"``` to overwrite an erroneous tagging of "greater than {self.lower_bound}" to (None,None)

I could also delete the vacuous case altogether so that if some programmer tried to create an Unbounded validator with the Bounded object they'd fail instantly. (with a "KeyError: (None,None) is not in this dict")

Reply:
i added `assert lower_bound is not None or upper_bound is not None, "Do not use this object to create an unbounded validator."` and removed the vaccuous case from the dict

Reply:
resolving this conv cuz the code is out of date